bootstrap-pull-request/src/prebuilt.ts

```diff
@@ -99,29 +91,80 @@ function assertIsPartialApplication(o: unknown): asserts o is PartialApplication
     assert(typeof o.metadata.annotations['github.head-sha'] === 'string', 'github.head-sha must be a string')
   }
+  assert('spec' in o, 'must have spec property')
+  assert(typeof o.spec === 'object', 'spec must be an object')
+  assert(o.spec !== null, 'spec must not be null')
+  assert('source' in o.spec, 'spec must have source property')
+  assert(typeof o.spec.source === 'object', 'source must be an object')
+  assert(o.spec.source !== null, 'source must not be null')
+  assert('path' in o.spec.source, 'source must have path property')
+  assert(typeof o.spec.source.path === 'string', 'path must be a string')
+}
+
+const parseApplicationManifest = async (applicationManifestPath: string): Promise<PartialApplication | Error> => {
+  let application
+  try {
+    application = yaml.load(await fs.readFile(applicationManifestPath, 'utf-8'))
+  } catch (error) {
+    if (error instanceof yaml.YAMLException) {
+      return error
+    }
+    throw error
+  }
+  try {
+    assertIsPartialApplication(application)
+  } catch (error) {
+    if (error instanceof assert.AssertionError) {
+      return error
+    }
+    throw error
+  }
+  return application
 }
 
 const writeServices = async (inputs: Inputs): Promise<void> => {
   const existingApplicationManifestPaths = await (
-    await glob.create(`${inputs.namespaceDirectory}/applications/**`, { matchDirectories: false })
+    await glob.create(`${inputs.namespaceDirectory}/applications/*.yaml`, { matchDirectories: false })
+  ).glob()
+  const prebuiltApplicationManifestPaths = await (
+    await glob.create(`${inputs.prebuiltDirectory}/applications/*.yaml`, { matchDirectories: false })
   ).glob()
+  for (const prebuiltApplicationManifestPath of prebuiltApplicationManifestPaths) {
+    const prebuiltApplication = await parseApplicationManifest(prebuiltApplicationManifestPath)
+    if (prebuiltApplication instanceof Error) {
+      const error: Error = prebuiltApplication
+      core.info(`Ignored an invalid application manifest: ${prebuiltApplicationManifestPath}: ${String(error)}`)
+      continue
+    }
 
-  core.info(`Finding services in ${inputs.prebuiltDirectory}/services`)
-  const services = (await fs.readdir(`${inputs.prebuiltDirectory}/services`, { withFileTypes: true }))
-    .filter((e) => e.isDirectory())
-    .map((e) => e.name)
-  for (const service of services) {
-    const applicationManifestPath = `${inputs.namespaceDirectory}/applications/${inputs.namespace}--${service}.yaml`
-    if (existingApplicationManifestPaths.includes(applicationManifestPath)) {
-      core.info(`Service ${service} is already pushed`)
+    if (!prebuiltApplication.spec.source.path.startsWith('services/')) {
+      core.info(`Ignored a non-service application manifest: ${prebuiltApplicationManifestPath}`)
       continue
     }
-    await writeServiceManifest(inputs, service)
-    await writeApplicationManifest(inputs, service)
+    const service = path.basename(prebuiltApplication.spec.source.path)
```

Comment:
Parse the service name from Application, like:
https://github.com/quipper/monorepo-deploy-actions/blob/4d5710442b0af5cf6030f5c538fa570e1ed7d115/bootstrap-pull-request/tests/fixtures/prebuilt/applications/prebuilt--a.yaml#L14-L17